src/rust/ide/lib/span-tree/src/node/kind.rs

```diff
@@ -0,0 +1,303 @@
+                true
+            },
+            _ => false
+        }
+    }
+}
+
+
+// === Impls ===
+
+impl Default for Kind {
+    fn default() -> Self {
+        Self::insertion_point().into()
+    }
+}
+
+
+
+// ============
+// === This ===
+// ============
+
+/// Kind representing "this" node. For example, in the following expressions, `foo` is considered
+/// "this": `foo.bar`, `foo + bar`, `foo.+ bar`.
+#[derive(Clone,Debug,Default,Eq,PartialEq)]
+#[allow(missing_docs)]
+pub struct This {
+    pub removable : bool,
+    pub tp        : Option<String>,
+}
+
+
+// === Smart Setters ===
+
+#[allow(missing_docs)]
+impl This {
+    pub fn removable      (mut self)                    -> Self { self.removable = true     ; self }
+    pub fn typed          (mut self, tp:String)         -> Self { self.tp        = Some(tp) ; self }
+    pub fn with_removable (mut self, removable:bool)    -> Self { self.removable = removable; self }
+    pub fn with_tp        (mut self, tp:Option<String>) -> Self { self.tp        = tp       ; self }
+}
+
+impl From<This> for Kind {
+    fn from(t:This) -> Self {
+        Self::This(t)
+    }
+}
+
+
+
+// ================
+// === Argument ===
+// ================
+
+/// Kind representing "argument" node. For example, in the following expressions, `a`, `b`, and `c`
+/// are considered "arguments": `foo a b c`, `foo + a`.
+#[derive(Clone,Default,Debug,Eq,PartialEq)]
+#[allow(missing_docs)]
+pub struct Argument {
+    pub removable : bool,
+    pub name      : Option<String>,
+    pub tp        : Option<String>,
+}
+
+
+// === Smart Setters ===
+
+#[allow(missing_docs)]
+impl Argument {
+    pub fn removable      (mut self)                      -> Self { self.removable = true  ; self }
+    pub fn named          (mut self, name:String)         -> Self { self.name = Some(name) ; self }
+    pub fn typed          (mut self, tp:String)           -> Self { self.tp = Some(tp)     ; self }
+    pub fn with_removable (mut self, rm:bool)             -> Self { self.removable = rm    ; self }
+    pub fn with_name      (mut self, name:Option<String>) -> Self { self.name = name       ; self }
+    pub fn with_tp        (mut self, tp:Option<String>)   -> Self { self.tp = tp           ; self }
+}
+
+impl From<Argument> for Kind {
+    fn from(t:Argument) -> Self {
+        Self::Argument(t)
+    }
+}
+
+
+
+// ======================
+// === InsertionPoint ===
+// ======================
+
+/// Kind representing "insertion point" node. For example, in the following expressions, all places
+/// marked with `?` are considered "insertion points" (please note that `?` symbols are NOT part
+/// of the source code): `foo ? a ? b ? c ? ? ?`. The final question marks indicate non-provided
+/// arguments.
+#[derive(Clone,Debug,Default,Eq,PartialEq)]
+#[allow(missing_docs)]
+pub struct InsertionPoint {
+    pub kind : InsertionPointType,
+    pub name : Option<String>,
+    pub tp   : Option<String>,
+}
+
+// === Smart Constructors ===
```

Comment:
I see everything here is smart :D And again it conflicts with some other programming idioms: https://wiki.haskell.org/Smart_constructors